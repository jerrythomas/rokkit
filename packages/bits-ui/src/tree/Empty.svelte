<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, ...rest } = $props()
</script>

<tree-empty {...rest} class={className} data-tree-empty>
	{@render children?.()}
</tree-empty>
