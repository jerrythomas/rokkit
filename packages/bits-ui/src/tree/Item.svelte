<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, ...rest } = $props()
</script>

<tree-node-content {...rest} class={className} data-tree-node-content>
	{@render children?.()}
</tree-node-content>
