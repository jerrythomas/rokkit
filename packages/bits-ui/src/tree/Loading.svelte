<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = 'm-auto p-4 text-center text-gray-500', children, ...rest } = $props()
</script>

<tree-loading {...rest} class={className} data-tree-loading>
	{@render children?.()}
</tree-loading>
