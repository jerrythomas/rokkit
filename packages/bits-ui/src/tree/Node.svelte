<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, ...rest } = $props()
</script>

<tree-node {...rest} class={className} data-tree-node>
	{@render children?.()}
</tree-node>
