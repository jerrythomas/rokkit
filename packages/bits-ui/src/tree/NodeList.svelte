<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, ...rest } = $props()
</script>

<tree-node-items {...rest} class={className} data-tree-node-items>
	{@render children?.()}
</tree-node-items>
