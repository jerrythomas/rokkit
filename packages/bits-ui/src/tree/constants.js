export const TREE_ROOT_ATTR = 'data-tree-root'
export const TREE_LOADING_ATTR = 'data-tree-loading'
export const TREE_EMPTY_ATTR = 'data-tree-empty'
export const TREE_VIEWPORT_ATTR = 'data-tree-viewport'

export const TREE_LIST_ATTR = 'data-tree-list'
export const TREE_ITEM_ATTR = 'data-tree-item'

export const TREE_VALUE_ATTR = 'data-value'
export const TREE_LINE_ATTR = 'data-tree-line'
export const TREE_LINE_TYPE_ATTR = 'data-tree-line-type'
export const TREE_PATH_ATTR = 'data-path'

// Export symbols for context
export const TREE_CTX = Symbol('tree')
export const TREE_LINE_TYPES = ['last', 'child', 'sibling', 'empty']
export const TREE_LINE_EMPTY = 'empty'
