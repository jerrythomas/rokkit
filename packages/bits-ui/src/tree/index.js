export { default as Root } from './Root.svelte'
export { default as Viewport } from './Viewport.svelte'
export { default as Node } from './Node.svelte'
export { default as NodeList } from './NodeList.svelte'
export { default as Item } from './Item.svelte'

export { default as Line } from './Line.svelte'
export { default as Empty } from './Empty.svelte'
export { default as Loading } from './Loading.svelte'
