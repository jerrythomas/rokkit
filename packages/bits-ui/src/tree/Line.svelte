<script>
	import { getContext } from 'svelte'
	import { TREE_CTX, TREE_LINE_TYPES, TREE_LINE_EMPTY } from './constants'

	const { rtl } = getContext(TREE_CTX)

	/**
	 * @typedef {Object} TreeLineProps
	 * @property {import('./types').TreeLineType} [type]
	 * @property {boolean} [rtl=false]
	 */

	/** @type {TreeLineProps} */
	let { type = TREE_LINE_EMPTY } = $props()

	let validatedType = $derived(TREE_LINE_TYPES.includes(type) ? type : TREE_LINE_EMPTY)
</script>

<tree-line
	class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
	data-tree-line
	data-tree-line-type={validatedType}
>
	{#if validatedType === 'last'}
		{#if rtl}
			<i class="border-b border-r"></i>
		{:else}
			<i class="border-r"></i>
			<i class="border-b"></i>
		{/if}
	{:else if validatedType === 'child'}
		{#if rtl}
			<i class="row-span-2 grid grid-rows-2 border-r">
				<i class="border-b"></i>
			</i>
		{:else}
			<i class="col-span-1 row-span-2 border-r"></i>
			<i class="border-b"></i>
		{/if}
	{:else if validatedType === 'sibling'}
		<i class="row-span-2 border-r"></i>
	{/if}
</tree-line>
