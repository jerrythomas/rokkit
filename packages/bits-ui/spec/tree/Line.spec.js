import { describe, expect, it, beforeEach, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import Line from '../../src/tree/Line.svelte'
import { TREE_CTX } from '../../src/tree/constants'

describe('Line.svelte', () => {
	const ROOT_SELECTOR = 'tree-line'
	beforeEach(() => {
		cleanup()
	})

	describe('given a type', () => {
		const context = new Map([[TREE_CTX, { rtl: false }]])

		describe('when type is not specified', () => {
			it('then it should render the correct elements', () => {
				const { container } = render(Line, { context })
				const node = container.querySelectorAll(ROOT_SELECTOR)
				expect(node).toHaveLength(1)
				expect(node[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
				expect(node[0].hasAttribute('data-tree-line')).toBeTruthy()
				expect(node[0].getAttribute('data-tree-line-type')).toEqual('empty')
				expect(node[0].children).toHaveLength(0)
			})
		})

		describe('when type is "empty"', () => {
			it('then it should render the correct elements', () => {
				const { container } = render(Line, { context, props: { type: 'empty' } })
				const span = container.querySelectorAll(ROOT_SELECTOR)
				expect(span).toHaveLength(1)
				expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
				expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
				expect(span[0].getAttribute('data-tree-line-type')).toEqual('empty')
				expect(container).toMatchSnapshot()
			})
		})

		describe('when type is "last"', () => {
			it('then it should render the correct elements', () => {
				const { container } = render(Line, { context, props: { type: 'last' } })
				const span = container.querySelectorAll(ROOT_SELECTOR)
				expect(span).toHaveLength(1)
				expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
				expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
				expect(span[0].getAttribute('data-tree-line-type')).toEqual('last')
				expect(container).toMatchSnapshot()
			})
		})

		describe('when type is "child"', () => {
			it('then it should render the correct elements', () => {
				const { container } = render(Line, { context, props: { type: 'child' } })
				const span = container.querySelectorAll(ROOT_SELECTOR)
				expect(span).toHaveLength(1)
				expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
				expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
				expect(span[0].getAttribute('data-tree-line-type')).toEqual('child')
				expect(span[0].children).toHaveLength(2)
				expect(container).toMatchSnapshot()
			})
		})

		describe('when type is "sibling"', () => {
			it('then it should render the correct elements', () => {
				const { container } = render(Line, { context, props: { type: 'sibling' } })

				const span = container.querySelectorAll(ROOT_SELECTOR)
				expect(span).toHaveLength(1)
				expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
				expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
				expect(span[0].getAttribute('data-tree-line-type')).toEqual('sibling')
				expect(container).toMatchSnapshot()
			})
		})
	})

	describe('When rtl is true property', () => {
		const context = new Map([[TREE_CTX, { rtl: true }]])

		beforeEach(() => {
			document.dir = 'rtl'
		})
		afterEach(() => {
			document.dir = 'ltr'
		})
		it('should render when type is "last"', () => {
			const { container } = render(Line, { context, props: { type: 'last' } })

			const span = container.querySelectorAll(ROOT_SELECTOR)
			expect(span).toHaveLength(1)
			expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
			expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
			expect(span[0].getAttribute('data-tree-line-type')).toEqual('last')
			expect(container).toMatchSnapshot()
		})

		it('should render when type is "child"', () => {
			const { container } = render(Line, { context, props: { type: 'child' } })

			const span = container.querySelectorAll(ROOT_SELECTOR)
			expect(span).toHaveLength(1)
			expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
			expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
			expect(span[0].getAttribute('data-tree-line-type')).toEqual('child')
			expect(container).toMatchSnapshot()
		})
		it('should render when type is "sibling"', () => {
			const { container } = render(Line, { context, props: { type: 'sibling' } })

			const span = container.querySelectorAll(ROOT_SELECTOR)
			expect(span).toHaveLength(1)
			expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
			expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
			expect(span[0].getAttribute('data-tree-line-type')).toEqual('sibling')
			expect(container).toMatchSnapshot()
		})
		it('should render when type is "empty"', () => {
			const { container } = render(Line, { context, props: { type: 'empty' } })

			const span = container.querySelectorAll(ROOT_SELECTOR)
			expect(span).toHaveLength(1)
			expect(span[0].className).toEqual('grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2')
			expect(span[0].hasAttribute('data-tree-line')).toBeTruthy()
			expect(span[0].getAttribute('data-tree-line-type')).toEqual('empty')
			expect(span[0].hasAttribute('data-line-last')).toBeFalsy()
			expect(container).toMatchSnapshot()
		})
	})
})
