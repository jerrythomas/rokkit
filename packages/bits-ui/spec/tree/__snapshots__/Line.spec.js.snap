// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Line.svelte > When rtl is true property > should render when type is "child" 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="child"
  >
    <!---->
    <i
      class="row-span-2 grid grid-rows-2 border-r"
    >
      <i
        class="border-b"
      />
    </i>
    
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > When rtl is true property > should render when type is "empty" 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="empty"
  >
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > When rtl is true property > should render when type is "last" 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="last"
  >
    <!---->
    <i
      class="border-b border-r"
    />
    
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > When rtl is true property > should render when type is "sibling" 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="sibling"
  >
    <i
      class="row-span-2 border-r"
    />
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > given a type > when type is "child" > then it should render the correct elements 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="child"
  >
    <!---->
    <i
      class="col-span-1 row-span-2 border-r"
    />
     
    <i
      class="border-b"
    />
    
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > given a type > when type is "empty" > then it should render the correct elements 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="empty"
  >
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > given a type > when type is "last" > then it should render the correct elements 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="last"
  >
    <!---->
    <i
      class="border-r"
    />
     
    <i
      class="border-b"
    />
    
    <!---->
  </tree-line>
  
</div>
`;

exports[`Line.svelte > given a type > when type is "sibling" > then it should render the correct elements 1`] = `
<div>
  <tree-line
    class="grid h-full w-4 min-w-4 grid-cols-2 grid-rows-2"
    data-tree-line="true"
    data-tree-line-type="sibling"
  >
    <i
      class="row-span-2 border-r"
    />
    <!---->
  </tree-line>
  
</div>
`;
